iModelCore/ECDb/ECDb/ECSql/ECSqlParser.cpp

```diff
@@ -3262,6 +3292,785 @@ BentleyStatus ECSqlParser::ParseValuesOrQuerySpec(std::vector<std::unique_ptr<Va
+    for (size_t nPos = 0; nPos < parseNode->count(); nPos++)
+        {
+        std::unique_ptr<WindowDefinitionExp> windowDefinitionExp = nullptr;
+        if (SUCCESS != ParseWindowDefinitionExp(windowDefinitionExp, parseNode->getChild(nPos)))
+            return ERROR;
+
+        windowDefinitionList.push_back(std::move(windowDefinitionExp));
+        }
+
+    windowDefinitionListExp = std::make_unique<WindowDefinitionListExp>(windowDefinitionList);
+    return SUCCESS;
+    }
+
+//-----------------------------------------------------------------------------------------
+// @bsimethod
+//+---------------+---------------+---------------+---------------+---------------+------
+BentleyStatus ECSqlParser::ParseWindowDefinitionExp(std::unique_ptr<WindowDefinitionExp>& windowDefinitionExp, OSQLParseNode const* parseNode) const
+    {
+    if (!SQL_ISRULE(parseNode, window_definition))
+        {
+        BeAssert(false && "Invalid grammar. Expecting window_definition");
+        return ERROR;
+        }
+
+    std::unique_ptr<WindowSpecification> windowSpecification = nullptr;
+    if (SUCCESS != ParseWindowSpecification(windowSpecification, parseNode->getChild(2)))
+        return ERROR;
+
+    windowDefinitionExp = std::make_unique<WindowDefinitionExp>(parseNode->getChild(0)->getTokenValue(), std::move(windowSpecification));
+    return SUCCESS;
+    }
+
+//-----------------------------------------------------------------------------------------
+// @bsimethod
+//+---------------+---------------+---------------+---------------+---------------+------
+BentleyStatus ECSqlParser::ParseWindowFunction(std::unique_ptr<ValueExp>& valueExp, OSQLParseNode const* parseNode) const
+    {
+    if (!SQL_ISRULE(parseNode, window_function))
+        {
+        BeAssert(false && "Invalid grammar. Expecting window_function");
+        return ERROR;
+        }
+
+    std::unique_ptr<ValueExp> functionCallExp = nullptr;
+
+    if (SUCCESS != ParseWindowFunctionType(functionCallExp, parseNode->getChild(0)))
+        return ERROR;
+
+    std::unique_ptr<FilterClauseExp> filterClauseExp = nullptr;
+    if (SUCCESS != ParseFilterClause(filterClauseExp, parseNode->getChild(1)))
+        return ERROR;
+
+    if (SQL_ISRULE(parseNode->getChild(3), window_specification))
+        {
+        std::unique_ptr<WindowSpecification> windowSpecificationExp = nullptr;
+        if (SUCCESS != ParseWindowSpecification(windowSpecificationExp, parseNode->getChild(3)))
+            return ERROR;
+
+        valueExp = std::make_unique<WindowFunctionExp>(std::move(functionCallExp), std::move(filterClauseExp), std::move(windowSpecificationExp));
+        return SUCCESS;
+        }
+    else if (parseNode->getChild(3)->getNodeType() == SQLNodeType::SQL_NODE_NAME)
+        {
+        Utf8CP windowName = parseNode->getChild(3)->getTokenValue().c_str();
+        valueExp = std::make_unique<WindowFunctionExp>(std::move(functionCallExp), std::move(filterClauseExp), std::move(windowName));

```

Comment:
The std::move on an Utf8CP seems excessive, we have to do it on the many unique_ptr in this file, but for a simple Utf8CP we should just pass by value.
Also, the getTokenValue() method returns an Utf8String, .c_str() gets its Utf8CP and the ctor of WindowFunctionExp takes an Utf8String as a parameter, so it will construct a copy of the string anyways. Better make the ctor take an Utf8CP instead and assign it to the Utf8String member, like LiteralValueExp does for example

Reply:
fixed